lib/std/zig/render.zig

```diff
@@ -15,54 +15,18 @@ pub const Error = error{
 pub fn render(allocator: *mem.Allocator, stream: var, tree: *ast.Tree) (@TypeOf(stream).Child.Error || Error)!bool {
     comptime assert(@typeInfo(@TypeOf(stream)) == .Pointer);
 
-    var anything_changed: bool = false;
-
     // make a passthrough stream that checks whether something changed
-    const MyStream = struct {
-        const MyStream = @This();
-        const StreamError = @TypeOf(stream).Child.Error;
-        const Stream = std.io.OutStream(StreamError);
-
-        anything_changed_ptr: *bool,
-        child_stream: @TypeOf(stream),
-        stream: Stream,
-        source_index: usize,
-        source: []const u8,
-
-        fn write(iface_stream: *Stream, bytes: []const u8) StreamError!usize {
-            const self = @fieldParentPtr(MyStream, "stream", iface_stream);
-
-            if (!self.anything_changed_ptr.*) {
-                const end = self.source_index + bytes.len;
-                if (end > self.source.len) {
-                    self.anything_changed_ptr.* = true;
-                } else {
-                    const src_slice = self.source[self.source_index..end];
-                    self.source_index += bytes.len;
-                    if (!mem.eql(u8, bytes, src_slice)) {
-                        self.anything_changed_ptr.* = true;
-                    }
-                }
-            }
+    const ChildStream = @TypeOf(stream).Child;
 
-            return self.child_stream.writeOnce(bytes);
-        }
-    };
-    var my_stream = MyStream{
-        .stream = MyStream.Stream{ .writeFn = MyStream.write },
-        .child_stream = stream,
-        .anything_changed_ptr = &anything_changed,
-        .source_index = 0,
-        .source = tree.source,
-    };
+    const ChangeDetectionStream = std.io.ChangeDetectionStream(ChildStream);
+    var change_detection_stream = ChangeDetectionStream.init(stream, tree.source);
 
-    try renderRoot(allocator, &my_stream.stream, tree);
+    const AutoIndentingStream = std.io.AutoIndentingStream(ChangeDetectionStream);
+    var auto_indenting_stream = AutoIndentingStream.init(&change_detection_stream, indent_delta);
```

Comment:
This is an example of being able to pipe streams together. Each stream is specialised on their down-stream type, their up-stream type can be anything (instead of being forced to used OutStream as the up-stream type).